docs/architecture/ErrorReporting.md

```diff
@@ -0,0 +1,144 @@
+# Handling and reporting errors
+
+    WARNING: This document describes future plans for error handling.
+    The UI library does not (yet) implement error handling as described here.
+
+This document describes how errors are handled within the ACS UI library and how they are reported to the surrounding application.
+
+## Error sources
+
+There are two categories of errors, and they are handled differently.
+
+1. ACS backend errors: Errors returned by the REST calls to ACS backend services.
+
+   1. These can be transient or permanent (special case: permission error from REST API).
+
+2. UI errors: Programming errors in the UI library / errors returned from other browser APIs. This is a catch-all for all errors not generated from the ACS SDK libraries.
+
+## ACS backend errors
+
+Handling ACS backend errors is challenging when using either the composites or the pure components with the react bindings provided by the ACS UI library:
+
+* ACS backend API calls are asynchronous
+* ACS backend API calls are often initiated outside of the React component lifecycle methods (e.g on interaction handler callbacks or `useEffect` blocks).
+
+But in both cases, all ACS backend API calls are mediated by the (chat and calling) stateful clients. Thus, the stateful client tees errors to a special field in the state:
+
+```ts
+interface CallClientState {
+    // ... other fields ...
+    errors: { [key: SdkClientMethod]: ErrorState[];
+}
+
+type SdkClientMethod = 'CallClient.createCallAgent' | 'CallClient.getDeviceManager'; // ... and many more
+
+interface ErrorState {
```

Comment:
@prprabhu-ms Have we looked at using a type similar to the standard error object defined for Microsoft REST APIs?
https://github.com/microsoft/api-guidelines/blob/vNext/Guidelines.md#error--object

```typescript
interface Error {
    code: string,
    message: string,
    target?: string,
    innererror?: Error;
    details?: Error[];
}
```

`target` can then be the function name

Reply:
Looked at this a bit and I think we'll have to stick to some very abstract type (perhaps any) for the error. Reasons:

* @azure/{communication-calling, communication-chat, communication-identity} do not return identical error types.
  * Chat returns the REST error (so mostly what you suggested).
  * Calling returns a different error type with code/subcode. Call transfer errors have their own shape.
  * Identity doesn't export any error types at all.
* We mostly get errors via rejected promises and typescript doesn't strongly type promise rejections: https://stackoverflow.com/questions/50071115/typescript-promise-rejection-type

So trying to strongly type the errors requires us to downcast and transform from `any` across a variety of types.

Will iterate on the exact type as I implement this for the stateful clients.